src/flag_gems/ops/weightnorm.py

```diff
@@ -0,0 +1,289 @@
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        w_value = tl.load(w + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        vw_block += v_value * w_value
+    vw_sum = tl.sum(vw_block, 1)[:, None]
+
+    for base in range(0, M, BLOCK_ROW_SIZE):
+        row_offset = base + ty
+        mask = row_offset < M and col_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        w_value = tl.load(w + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        v_grad_value = g_value * (
+            w_value / (norm_value + eps)
+            - v_value / (norm_value * norm_value * norm_value + eps) * vw_sum
+        )
+        tl.store(v_grad + row_offset * N + col_offset, v_grad_value, mask=mask)
+
+    g_grad_value = vw_sum / (norm_value + eps)
+    tl.store(g_grad + col_offset, g_grad_value, mask=col_mask)
+
+
+@libentry()
+@triton.autotune(configs=cfggen_first(), key=["M", "N"])
+@triton.jit(do_not_specialize=["eps"])
+def weight_norm_bwd_kernel_first(
+    v_grad,
+    g_grad,
+    w,
+    v,
+    g,
+    norm,
+    M,
+    N,
+    eps: tl.constexpr,
+    BLOCK_ROW_SIZE: tl.constexpr,
+    BLOCK_COL_SIZE: tl.constexpr,
+):
+    ty = tl.arange(0, BLOCK_ROW_SIZE)[:, None]
+    by = tl.program_id(axis=0) * BLOCK_ROW_SIZE
+    row_offset = by + ty
+    row_mask = row_offset < M
+
+    g_value = tl.load(g + row_offset, mask=row_mask).to(tl.float32)
+    norm_value = tl.load(norm + row_offset, mask=row_mask).to(tl.float32)
+
+    tx = tl.arange(0, BLOCK_COL_SIZE)[None, :]
+
+    v_block = tl.zeros([BLOCK_ROW_SIZE, BLOCK_COL_SIZE], dtype=tl.float32)
+    for base in range(0, N, BLOCK_COL_SIZE):
+        col_offset = base + tx
+        mask = col_offset < N and row_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        w_value = tl.load(w + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        v_block += v_value * w_value
+    vw_sum = tl.sum(v_block, 1)[:, None]
+
+    for base in range(0, N, BLOCK_COL_SIZE):
+        col_offset = base + tx
+        mask = col_offset < N and row_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        w_value = tl.load(w + row_offset * N + col_offset, mask=mask).to(tl.float32)
+        v_grad_value = g_value * (
+            w_value / (norm_value + eps)
+            - v_value / (norm_value * norm_value * norm_value + eps) * vw_sum
+        )
+        tl.store(v_grad + row_offset * N + col_offset, v_grad_value, mask=mask)
+
+    g_grad_value = vw_sum / (norm_value + eps)
+    tl.store(g_grad + row_offset, g_grad_value, mask=row_mask)
+
+
+class WeightNorm(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, v, g, dim):
+        logging.debug("GEMS WEIGHTNORM FORWARD")
+        v = v.contiguous()
+        g = g.contiguous()
+        output = torch.empty_like(v)
+        norm = torch.empty_like(g, dtype=torch.float32)
```

Comment:
`dtype=torch.float32)` here will cause a dtype dismatch error in my test environment. Does it work correctly in yours?

Reply:
i forget push a fix in `test_norm_ops.py`, is your error occur in this file?

Reply:
> i forget push a fix in `test_norm_ops.py` and have pushed now, is your error occur in this file?

Although `gems_assert_close(..., res_norm_out.dtype)` passed, but I think we should align the output dtype of res and ref instead of modifying the test code.  If we do need to alloc float32 tensor here, maybe we can `return output, norm.to(g.dtype)` at the end of this function, or even `norm = norm.to(g.dtype) before save ctx for bwd`.